core/chains/evm/logpoller/log_poller.go

```diff
@@ -896,12 +896,19 @@ func (lp *logPoller) blocksFromFinalizedLogs(ctx context.Context, logs []types.L
 // or if there is an error backfilling.
 func (lp *logPoller) backfill(ctx context.Context, start, end int64) error {
 	batchSize := lp.backfillBatchSize
+	stopCtx, cancel := lp.stopCh.NewCtx()
+	defer cancel()
+
 	for from := start; from <= end; from += batchSize {
 		to := mathutil.Min(from+batchSize-1, end)
 
 		gethLogs, err := lp.ec.FilterLogs(ctx, lp.Filter(big.NewInt(from), big.NewInt(to), nil))
 		if err != nil {
-			if !client.IsTooManyResults(err, lp.clientErrors) {
+			if stopCtx.Err() != nil {
+				return err
+			}
+			// Retry on context deadline exceeded since we may be able to get logs by reducing the block range.
+			if !client.IsTooManyResults(err, lp.clientErrors) && !errors.Is(err, context.DeadlineExceeded) {
```

Comment:
Nit: should not DeadlineExceeded handling be part of the IsTooManyResults logic.
Also could you add a test?

Reply:
I don't feel strongly about it either way, but looks like we do do something similar with `IsTimeout` for send errors:
```
// IsTimeout indicates if the error was caused by an exceeded context deadline
func (s *SendError) IsTimeout() bool {
    if s == nil {
        return false
    }
    if s.err == nil {
        return false
    }
    return pkgerrors.Is(s.err, context.DeadlineExceeded)
}
```